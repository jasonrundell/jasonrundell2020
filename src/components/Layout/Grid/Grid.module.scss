@import '../../../../src/styles/main.scss';

.root {
  column-count: 1;
  width: 100%;
}

@for $i from 1 through 12 {
  .col--#{$i} {
    columns: $i;
  }
  .col--medium--#{$i} {
    @media (min-width: $mediumBreakpoint) {
      columns: $i;
    }
  }
  .col--large--#{$i} {
    @media (min-width: $largeBreakpoint) {
      columns: $i;
    }
  }
}

.break-inside__auto {
  /** TODO: Fix the way we use break-inside. Should not use element selectors **/
  article,
  div,
  section {
    break-inside: auto;
  }
}

.break-inside__avoid {
  /** TODO: Fix the way we use break-inside. Should not use element selectors **/
  article,
  div,
  section {
    break-inside: avoid;
  }
}
