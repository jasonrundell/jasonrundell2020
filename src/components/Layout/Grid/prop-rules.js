const columnRules = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
const breakInsideRules = [
  'auto',
  'avoid',
  'avoid-page',
  'avoid-column',
  'avoid-region',
  'inherit',
  'initial',
  'unset',
]
export { columnRules, breakInsideRules }
