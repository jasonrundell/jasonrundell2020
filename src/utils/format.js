/**
 *  Letter case conversion functions
 *  [Letter case](https://en.wikipedia.org/wiki/Letter_case#Special_case_styles)
 *
 *  Converts only letter case-separated or hyphen-delimited strings
 */

/**
 *  Converts hyphen-delimited `string` to camel case
 *
 *  @example
 *  'foo-bar' => 'fooBar'
 *  'foo--bar' => 'fooBar'
 *
 *  @param {string} string
 */
function toCamelCase(string) {
  return string.replace(/(-+)([a-z])/g, ($0, $1, $2) => $2.toUpperCase())
}

/**
 *  Converts hyphen-delimited `string` to pascal case (aka upper camel case)
 *
 *  @example
 *  'foo-bar' => 'FooBar'
 *  'foo--bar' => 'FooBar'
 *
 *  @param {string} string
 */
function toPascalCase(string) {
  return toCamelCase(string.replace(/^[a-z]/, $0 => $0.toUpperCase()))
}

export { toCamelCase, toPascalCase }
