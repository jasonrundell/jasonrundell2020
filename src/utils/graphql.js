export function edgesToObject(edgesData) {
  const nodes = edgesData.map(item => {
    return item.node
  })
  const finalObject = nodes.map(item => {
    return item
  })

  return finalObject
}
