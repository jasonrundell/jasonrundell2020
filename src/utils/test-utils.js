export function catchWarnings() {
  const originalConsoleError = global.console.error
  beforeEach(() => {
    /**
     * I Want to know if values passed to props fail the type check
     * */
    global.console.error = (...args) => {
      const propTypeFailures = [/Warning:/] // e.g. [/Invalid prop/, /Failed prop type/]

      if (propTypeFailures.some(p => p.test(args[0]))) {
        throw new Error(args[0])
      }

      originalConsoleError(...args)
    }
  })
}

/**
 * Mix an match help from:
 * https://daveceddia.com/testing-for-proptypes-errors/
 * https://www.jackfranklin.co.uk/blog/failing-tests-on-react-proptypes/
 */
