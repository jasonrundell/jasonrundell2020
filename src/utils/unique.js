const onlyUnique = (value, index, self) => {
  return self.indexOf(value) === index
}

const generatePassword = ({
  passwordLength = 16,
  charset = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*()_-=~`',
}) => {
  let retVal = ''
  for (var i = 0, n = charset.length; i < passwordLength; ++i) {
    retVal += charset.charAt(Math.floor(Math.random() * n))
  }
  return retVal
}

export { onlyUnique, generatePassword }
