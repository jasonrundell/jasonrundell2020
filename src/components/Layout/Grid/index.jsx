import React from 'react'
import PropTypes from 'prop-types'

import { classNames } from '../../../utils/css'

import styles from './Grid.module.scss'

const columns = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]

const Grid = ({
  columnCount,
  mediumColumnCount,
  largeColumnCount,
  breakInside,
  children,
}) => {
  const colClassName = `col${columnCount}`
  const mediumColClassName = `colMedium${mediumColumnCount}`
  const largeColClassName = `colLarge${largeColumnCount}`
  const breakInsideClassName = `breakInside__${breakInside}`

  const className = classNames(
    styles.root,
    styles[colClassName],
    styles[mediumColClassName],
    styles[largeColClassName],
    breakInside && styles[breakInsideClassName]
  )

  return <div className={className}>{children}</div>
}

Grid.defaultProps = {
  columnCount: 1,
  mediumColumnCount: 1,
  largeColumnCount: 1,
}

Grid.propTypes = {
  columnCount: PropTypes.oneOf(columns),
  mediumColumnCount: PropTypes.oneOf(columns),
  largeColumnCount: PropTypes.oneOf(columns),
  breakInside: PropTypes.oneOf([
    'auto',
    'avoid',
    'avoid-page',
    'avoid-column',
    'avoid-region',
    'inherit',
    'initial',
    'unset',
  ]),
  children: PropTypes.any.isRequired,
}

export default Grid
