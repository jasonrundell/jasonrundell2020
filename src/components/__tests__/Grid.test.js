import React from 'react'
import { render, unmountComponentAtNode } from 'react-dom'
import { act } from 'react-dom/test-utils'
import renderer from 'react-test-renderer'
import { catchWarnings } from '../../utils/test-utils'

import Grid from '../Layout/Grid'
import Rules from '../Layout/Grid/prop-rules'

const { columnRules, breakInsideRules } = Rules

catchWarnings()

let container = null
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  unmountComponentAtNode(container)
  container.remove()
  container = null
})

describe('Grid component', () => {
  it('Grid renders correctly', () => {
    const tree = renderer
      .create(
        <div>
          <Grid>
            <div>Item</div>
            <div>Item</div>
            <div>Item</div>
          </Grid>
          <Grid columnCount={columnRules[0]}>
            <div>Item</div>
            <div>Item</div>
            <div>Item</div>
          </Grid>
          <Grid columnCount={columnRules[columnRules.length - 1]}>
            <div>Item</div>
            <div>Item</div>
            <div>Item</div>
          </Grid>
          <Grid mediumColumnCount={columnRules[0]}>
            <div>Item</div>
            <div>Item</div>
            <div>Item</div>
          </Grid>
          <Grid mediumColumnCount={columnRules[columnRules.length - 1]}>
            <div>Item</div>
            <div>Item</div>
            <div>Item</div>
          </Grid>
          <Grid largeColumnCount={columnRules[0]}>
            <div>Item</div>
            <div>Item</div>
            <div>Item</div>
          </Grid>
          <Grid largeColumnCount={columnRules[columnRules.length - 1]}>
            <div>Item</div>
            <div>Item</div>
            <div>Item</div>
          </Grid>
          <Grid breakInside={breakInsideRules[0]}>
            <div>Item</div>
            <div>Item</div>
            <div>Item</div>
          </Grid>
          <Grid breakInside={breakInsideRules[breakInsideRules.length - 1]}>
            <div>Item</div>
            <div>Item</div>
            <div>Item</div>
          </Grid>
        </div>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Grid has required children prop', () => {
    act(() => {
      render(<Grid>Grid component</Grid>, container)
    })
    expect(container.textContent).toBe('Grid component')
  })
})
