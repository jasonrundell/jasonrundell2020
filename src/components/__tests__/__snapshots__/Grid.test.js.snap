// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid component Grid renders correctly 1`] = `
<div>
  <div
    className="root col1 colMedium1 colLarge1"
  >
    <div>
      Item
    </div>
    <div>
      Item
    </div>
    <div>
      Item
    </div>
  </div>
  <div
    className="root col1 colMedium1 colLarge1"
  >
    <div>
      Item
    </div>
    <div>
      Item
    </div>
    <div>
      Item
    </div>
  </div>
  <div
    className="root col1 colMedium1 colLarge1"
  >
    <div>
      Item
    </div>
    <div>
      Item
    </div>
    <div>
      Item
    </div>
  </div>
  <div
    className="root col1 colMedium1 colLarge1"
  >
    <div>
      Item
    </div>
    <div>
      Item
    </div>
    <div>
      Item
    </div>
  </div>
  <div
    className="root col1 colMedium1 colLarge1"
  >
    <div>
      Item
    </div>
    <div>
      Item
    </div>
    <div>
      Item
    </div>
  </div>
  <div
    className="root col1 colMedium1 colLarge1"
  >
    <div>
      Item
    </div>
    <div>
      Item
    </div>
    <div>
      Item
    </div>
  </div>
  <div
    className="root col1 colMedium1 colLarge1"
  >
    <div>
      Item
    </div>
    <div>
      Item
    </div>
    <div>
      Item
    </div>
  </div>
  <div
    className="root col1 colMedium1 colLarge1 breakInside__auto"
  >
    <div>
      Item
    </div>
    <div>
      Item
    </div>
    <div>
      Item
    </div>
  </div>
  <div
    className="root col1 colMedium1 colLarge1"
  >
    <div>
      Item
    </div>
    <div>
      Item
    </div>
    <div>
      Item
    </div>
  </div>
</div>
`;
